### ISSUE TITLE
HorizontalPodAutoScaler with VictoriaMetrics operator

### ISSUE THREAD

**Issue by mshalenko** [VictoriaMetrics/operator#247]
Hi, everybody!

For example, current replicaCount of Deployment vminsert-cluster is 7 and if I create the HorizontalPodAutoscaler (then HPA) as shown below, then at low utilization, HPA will start to terminate pods, and the operator will start re-creating them to 7 pods as stated in the CustomResourceDefinition (then CRD) VMCluster.

```
apiVersion: autoscaling/v2beta2
kind: HorizontalPodAutoscaler
metadata:
  name: vminsert-cluster
  namespace: vm-monitoring
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: vminsert-cluster
  minReplicas: 3
  maxReplicas: 10
  metrics:
    - type: Resource
      resource:
        name: cpu
        target:
          averageUtilization: 80
          type: Utilization
```

Is it possible to use HPA with CRD of Victoria Metrics Operator?

**Comment by f41gh7** [comment]
Hello, thanks for reporting! For now its not possible. 

 But i think, its not hard to implement.

  My vision - introduce new config option `hpa` and `vpa`. with following syntax:

```yaml
kind: VMCluster
spec:
  vmselect:
      hpa:
       max: 3
       min: 1
       metrics:
              - type: Resource
      resource:
        name: cpu
        target:
          averageUtilization: 80
          type: Utilization
  vminsert:
      hpa:
       max: 3
       min: 1
       metrics:
              - type: Resource
      resource:
        name: cpu
        target:
          averageUtilization: 80
          type: Utilization
```

 This configuration will produce two HPA objects with needed config.

**Comment by f41gh7** [comment]
Master branch contains HPA settings for vminsert and vmselect.

 It can be used with following syntax:

```yaml
kind: VMCluster
spec:
  vmselect:
      hpa:
       max: 3
       min: 1
       metrics:
       - type: Resource
         resource:
            name: cpu
            target:
              averageUtilization: 80
              type: Utilization
  vminsert:
      hpa:
       max: 3
       min: 1
       metrics:
       - type: Resource
         resource:
            name: cpu
            target:
              averageUtilization: 80
              type: Utilization
```
 API doc https://github.com/VictoriaMetrics/operator/blob/master/docs/api.MD#embeddedhpa

 There is one issue with vmselect auto-scaling - its needed to perform update for all vmselect nodes. Because vmselect configuration contains list of all vmselect nodes.

   Also, i think HPA is reasonable, when you have different load during some time and you want to reduce costs. But i'm afraid, that in most cases  `vmstorage` will be a bottleneck and there is no need in HPA.

 Can you share your experience with vmcluster HPA later? Its very interesting.

**Comment by mshalenko** [comment]
@f41gh7 Thanks, I'll come back later with news.

**Comment by mshalenko** [comment]
@f41gh7 How can I deploy vm-operator with this changes? 1) Using docker hub image with tag latest? 2) Or building image manually? If second way is true, could you tell me in more detail how to do it correctly?

**Comment by f41gh7** [comment]
> @f41gh7 How can I deploy vm-operator with this changes? 1) Using docker hub image with tag latest? 2) Or building image manually? If second way is true, could you tell me in more detail how to do it correctly?

 I suggest to use branch based image - victoriametrics/operator:master

**Comment by mshalenko** [comment]
@f41gh7 I builded image using `make build`.
I deployed VM with hpa changes, and I tested it with vminsert. It looks nice. We can close the task. Thanks

**Comment by f41gh7** [comment]
Thanks! Any performance benefits  with auto-scaling of vminsert?

 Changes was included to the `0.15.0` release.
